File: dotnet/aws/Mcma.Aws.DynamoDb/DynamoDbTable.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;
using Amazon.DynamoDBv2;
using Amazon.DynamoDBv2.DocumentModel;
using Mcma.Data;
using Mcma.Data.DocumentDatabase.Queries;
using Mcma.Serialization;
using Newtonsoft.Json.Linq;

namespace Mcma.Aws.DynamoDb
{
    public class DynamoDbTable : IDocumentDatabaseTable
    {
        public DynamoDbTable(IAmazonDynamoDB dynamoDb, DynamoDbTableDescription tableDescription, DynamoDbTableOptions options = null)
        {
            TableDescription = tableDescription ?? throw new ArgumentNullException(nameof(tableDescription));
            Options = options ?? new DynamoDbTableOptions();
            Table = dynamoDb != null ? Table.LoadTable(dynamoDb, tableDescription.TableName) : throw new ArgumentNullException(nameof(dynamoDb));
        }

        private DynamoDbTableDescription TableDescription { get; }

        private DynamoDbTableOptions Options { get; }

        private Table Table { get; }

        private static (string partitionKey, string sortKey) ParsePartitionAndSortKeys(string id)
        {
            var lastSlashIndex = id.LastIndexOf('/');
            return lastSlashIndex > 0
                       ? (id.Substring(0, lastSlashIndex), id.Substring(lastSlashIndex + 1))
                       : (id, id);
        }

        private static T DocumentToResource<T>(Document document) where T : class
        {
            var docJson = JObject.Parse(document.ToJson());

            var resourceJson = docJson["resource"];

            return resourceJson?.ToMcmaObject<T>();
        }

        private Document ResourceToDocument<T>(string partitionKey, string sortKey, T resource)
        {
            var resourceJson = resource.ToMcmaJson().RemoveEmptyStrings();
            
            var item = new JObject
            {
                [TableDescription.PartitionKeyName] = partitionKey,
                [TableDescription.SortKeyName] = sortKey,
                [nameof(resource)] = resourceJson
            };

            foreach (var kvp in Options.TopLevelAttributeMappings)
                item[kvp.Key] = resourceJson.SelectToken(kvp.Value);

            return Document.FromJson(item.ToString());
        }

        private static async Task<IEnumerable<T>> ExecuteScanOrQueryAsync<T>(Func<Search> executeQueryOrScan, int? pageSize, int? pageNumber) where T : class
        {
            var items = new List<T>();
            var itemsReturned = 0;
            
            var search = executeQueryOrScan();
            do
            {
                var data = await search.GetNextSetAsync();

                var itemsToAdd = new List<Document>();
                if (!pageSize.HasValue)
                {
                    itemsToAdd.AddRange(data);
                }
                else
                {
                    var startIndex = pageSize.Value * pageNumber.GetValueOrDefault();
                    var endIndex = startIndex + pageSize.Value;

                    for (var i = 0; i < data.Count; i++)
                    {
                        var currentOverallIndex = itemsReturned + i;
                        if (currentOverallIndex < startIndex || currentOverallIndex >= endIndex)
                            continue;
                        
                        itemsToAdd.Add(data[i]);

                        if (items.Count + itemsToAdd.Count == pageSize.Value)
                            break;
                    }

                    items.AddRange(itemsToAdd.Select(DocumentToResource<T>));

                    itemsReturned += data.Count;
                }
            }
            while (!search.IsDone && (!pageSize.HasValue || items.Count < pageSize));

            return items;
        }

        private async Task<IEnumerable<T>> ExecuteQueryAsync<T>(Query<T> query) where T : class
        {
            var queryOpConfig = new QueryOperationConfig
            {
                KeyExpression = new Expression
                {
                    ExpressionStatement = "#partitionKey = :partitionKey",
                    ExpressionAttributeNames = new Dictionary<string, string> {["#partitionKey"] = TableDescription.PartitionKeyName},
                    ExpressionAttributeValues = new Dictionary<string, DynamoDBEntry> {[":partitionKey"] = query.Path}
                },
                FilterExpression = DynamoDbFilterExpressionBuilder.Build(query.FilterExpression),
                ConsistentRead = Options.ConsistentQuery ?? false
            };

            if (!string.IsNullOrWhiteSpace(query.SortBy))
            {
                queryOpConfig.IndexName = GetIndexName(query.SortBy);
                queryOpConfig.BackwardSearch = !query.SortAscending;
            }

            return await ExecuteScanOrQueryAsync<T>(() => Table.Query(queryOpConfig),
                                                    query.PageSize,
                                                    query.PageNumber);
        }

        private async Task<IEnumerable<T>> ExecuteScanAsync<T>(Query<T> query) where T : class
        {
            var scanOpConfig = new ScanOperationConfig
            {
                FilterExpression = DynamoDbFilterExpressionBuilder.Build(query.FilterExpression),
                ConsistentRead = Options.ConsistentQuery ?? false
            };

            if (!string.IsNullOrWhiteSpace(query.SortBy))
            {
                if (!query.SortAscending)
                    throw new McmaException("Scan does not support descending sorts.");
                scanOpConfig.IndexName = GetIndexName(query.SortBy);
            }

            return await ExecuteScanOrQueryAsync<T>(() => Table.Scan(scanOpConfig),
                                                    query.PageSize,
                                                    query.PageNumber);
        }

        private string GetIndexName(string sortBy)
        {
            var indexName = Options != null && Options.IndexNameMappings.ContainsKey(sortBy) ? Options.IndexNameMappings[sortBy] : sortBy;
            
            if (!TableDescription.IndexNames.Contains(indexName))
                throw new McmaException($"Invalid sortBy '{sortBy}'. A matching index was not found for the table.");
            
            return indexName;
        }

        public async Task<IEnumerable<T>> QueryAsync<T>(Query<T> query) where T : class
        {
            return await ExecuteQueryAsync(query);
        }

        public async Task<T> GetAsync<T>(string id) where T : class
        {
            var (partitionKey, sortKey) = ParsePartitionAndSortKeys(id);

            var document = await Table.GetItemAsync(partitionKey,
                                                    sortKey,
                                                    new GetItemOperationConfig {ConsistentRead = Options.ConsistentGet ?? false});

            return document != null ? DocumentToResource<T>(document) : null;
        }

        public async Task<T> PutAsync<T>(string id, T resource) where T : class
        {
            var (partitionKey, sortKey) = ParsePartitionAndSortKeys(id);
            await Table.PutItemAsync(ResourceToDocument(partitionKey, sortKey, resource));
            return resource;
        }

        public async Task DeleteAsync(string id)
        {
            var (partitionKey, sortKey) = ParsePartitionAndSortKeys(id);
            await Table.DeleteItemAsync(partitionKey, sortKey);
        }

        public IDocumentDatabaseMutex CreateMutex(string mutexName, string mutexHolder, TimeSpan? lockTimeout)
            => new DynamoDbMutex(Table, TableDescription, mutexName, mutexHolder, lockTimeout);
    }
}

File: nodejs/aws/dynamodb/lib/dynamo-db-table.ts
import { DynamoDB } from "aws-sdk";
import { McmaException, Utils } from "@mcma/core";
import { DocumentDatabaseTable, Query, Document } from "@mcma/data";
import { types } from "util";

import { DynamoDbTableOptions } from "./dynamo-db-table-options";
import { DynamoDbQueryFilter } from "./dynamo-db-query-filter";
import { DynamoDbTableDescription } from "./dynamo-db-table-description";
import { DocumentClient } from "aws-sdk/clients/dynamodb";

function parsePartitionAndSortKeys(id: string): { partitionKey: string, sortKey: string } {
    const lastSlashIndex = id.lastIndexOf("/");
    return lastSlashIndex > 0
        ? { partitionKey: id.substr(0, lastSlashIndex) , sortKey: id.substr(lastSlashIndex + 1) }
        : { partitionKey: id, sortKey: id };
}

export class DynamoDbTable implements DocumentDatabaseTable {
    private docClient: DocumentClient;
    
    constructor(
        dynamoDb: DynamoDB,
        private tableDescription: DynamoDbTableDescription,
        private options?: DynamoDbTableOptions
    ) {
        this.docClient = new DocumentClient({ service: dynamoDb });
    }

    private serialize(object: any) {
        if (object) {
            for (const key of Object.keys(object)) {
                const value = object[key];
                if (types.isDate(value)) {
                    if (isNaN(value.getTime())) {
                        delete object[key];
                    } else {
                        object[key] = value.toISOString();
                    }
                } else if (typeof value === "object") {
                    object[key] = this.serialize(value);
                }
            }
        }
        return object;
    }

    private deserialize(object: any) {
        if (object) {
            for (const key of Object.keys(object)) {
                const value = object[key];
                if (Utils.isValidDateString(value)) {
                    object[key] = new Date(value);
                } else if (typeof value === "object") {
                    object[key] = this.deserialize(value);
                }
            }
        }
        return object;
    }

    private async executeScanOrQuery<TDocument extends Document = Document> (
        executeScanOrQuery: (() => Promise<DocumentClient.ScanOutput> | Promise<DocumentClient.QueryOutput>),
        params: DocumentClient.ScanInput | DocumentClient.QueryInput,
        pageSize?: number,
        pageNumber?: number
    ): Promise<TDocument[]> {
        const items = [];
        let itemsReturned = 0;
        do
        {
            const data = await executeScanOrQuery();
            
            let itemsToAdd: DocumentClient.ItemList = [];
            if (!pageSize) {
                itemsToAdd.push(...data.Items);
            } else {
                // calculate the start and end index for the page we're looking for
                const startIndex = pageSize * (pageNumber ?? 0);
                const endIndex = startIndex + pageSize;

                for (let i = 0; i < data.Count; i++) {
                    const currentOverallIndex = itemsReturned + i;
                    if (currentOverallIndex >= startIndex && currentOverallIndex < endIndex) {
                        // add the item
                        itemsToAdd.push(data.Items[i]);

                        // if we reached the page size, we can break out now
                        if (items.length + itemsToAdd.length == pageSize) {
                            break;
                        }
                    } 
                }
            }

            params.ExclusiveStartKey = data.LastEvaluatedKey;

            // deserialize and add to master list 
            items.push(...itemsToAdd.map(i => this.deserialize(i.resource)));

            // track the total number of returned items
            itemsReturned += data.Count;
        }
        while (params.ExclusiveStartKey && (!pageSize || items.length < pageSize));

        return items;
    }

    private async executeQuery<TDocument extends Document = Document>(query: Query<TDocument>): Promise<TDocument[]> {
        let keyConditionExpression = "#partitionKey = :partitionKey";
        let expressionAttributeNames: DocumentClient.ExpressionAttributeNameMap = { "#partitionKey": this.tableDescription.partitionKeyName };
        let expressionAttributeValues: DocumentClient.ExpressionAttributeValueMap = { ":partitionKey": query.path };

        let filterExpression: string;
        if (query.filterExpression) {
            const dynamoDbFilter = new DynamoDbQueryFilter(query.filterExpression);
            dynamoDbFilter.build();
            
            filterExpression = dynamoDbFilter.expression;
            expressionAttributeNames = Object.assign(expressionAttributeNames, dynamoDbFilter.attributeNames);
            expressionAttributeValues = Object.assign(expressionAttributeValues, dynamoDbFilter.attributeValues);
        }

        const params: DocumentClient.QueryInput = {
            TableName: this.tableDescription.tableName,
            KeyConditionExpression: keyConditionExpression,
            FilterExpression: filterExpression,
            ExpressionAttributeNames: expressionAttributeNames,
            ExpressionAttributeValues: expressionAttributeValues,
            ConsistentRead: this.options?.consistentQuery
        };

        return this.executeScanOrQuery(
            () => this.docClient.query(params).promise(),
            params,
            query.pageSize,
            query.pageNumber
        );
    }

    private async executeScan<TDocument extends Document = Document>(query: Query<TDocument>): Promise<TDocument[]> {
        const dynamoDbFilter = new DynamoDbQueryFilter(query.filterExpression);
        dynamoDbFilter.build();

        const params: DocumentClient.ScanInput = {
            TableName: this.tableDescription.tableName,
            FilterExpression: dynamoDbFilter.expression,
            ExpressionAttributeNames:  dynamoDbFilter.attributeNames,
            ExpressionAttributeValues: dynamoDbFilter.attributeValues,
            ConsistentRead: this.options?.consistentQuery
        };

        return this.executeScanOrQuery(
            () => this.docClient.scan(params).promise(),
            params,
            query.pageSize,
            query.pageNumber
        );
    }

    async query<TDocument extends Document = Document>(query: Query<TDocument>): Promise<TDocument[]> {
        return this.executeQuery(query);
    }
    
    async get<TDocument extends Document = Document>(id: string): Promise<TDocument> {
        const { partitionKey, sortKey } = parsePartitionAndSortKeys(id);
        const params: DocumentClient.GetItemInput = {
            TableName: this.tableDescription.tableName,
            Key: {
                [this.tableDescription.partitionKeyName]: partitionKey,
                [this.tableDescription.sortKeyName]: sortKey
            },
            ConsistentRead: this.options?.consistentGet
        };

        const data = await this.docClient.get(params).promise();
        
        return data?.Item?.resource ? this.deserialize(data.Item.resource) : null;
    }

    async put<TDocument extends Document = Document>(id: string, resource: TDocument): Promise<TDocument> {
        const { partitionKey, sortKey } = parsePartitionAndSortKeys(id);
        resource = this.serialize(resource);

        const params: DocumentClient.PutItemInput = {
            TableName: this.tableDescription.tableName,
            Item: {
                [this.tableDescription.partitionKeyName]: partitionKey,
                [this.tableDescription.sortKeyName]: sortKey,
                resource
            }
        };
        try {
            await this.docClient.put(params).promise();
        } catch (error) {
            throw new McmaException("Failed to put resource in DynamoDB table", error);
        }
        return this.deserialize(resource);
    }

    async delete(id: string): Promise<void> {
        const { partitionKey, sortKey } = parsePartitionAndSortKeys(id);
        const params: DocumentClient.DeleteItemInput = {
            TableName: this.tableDescription.tableName,
            Key: {
                [this.tableDescription.partitionKeyName]: partitionKey,
                [this.tableDescription.sortKeyName]: sortKey
            }
        };
        try {
            await this.docClient.delete(params).promise();
        } catch (error) {
            throw new McmaException("Failed to delete resource in DynamoDB table", error);
        }
    }
}
